pyannote/database/filelist.py

```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env python
+# encoding: utf-8
+
+# The MIT License (MIT)
+
+# Copyright (c) 2016 CNRS
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# AUTHORS
+# Hervé BREDIN - http://herve.niderb.fr
+
+from pyannote.core import Segment, Timeline, Annotation
+from .database import Database
+from . import protocol as Protocol
+import functools
+from pathlib import Path
+import yaml
+import pandas as pd
+
+
+def subset_iter(database_name, path_mdtm=None, path_uem=None):
```

Comment:
We should ask for a mandatory `path_lst` argument as well that contains one uri per line.
This will add support for files with no annotation (which would otherwise be ignored).

`path_rttm` and `path_uem` can then become really optional.
When `path_rttm` is missing, we don't add the "annotation" key.
When `path_uem` is missing, we don't add the "annotated" key.

Reply:
I think `path_mdtm` has to be changed to a more generic `path_annotations` and we can allow it to be either in `mdtm`, `rttm`, or a simple `list` format (this one with extension `.lst`). So, we parse it according to its extension. So there will be only 2 path parameters and path_uem can be optional.

But I wonder, will pyannote work correctly if `uri` is yielded with both annotations and annotated keys empty? I assumed a pyannote database has to have some sort of annotations, even if the timeline covers the whole file.

Reply:
The problem with allowing to pass `mdtm` but not `lst` is that there is no way to know that a file with no annotation exists (because it does not appear in the `mdtm` file).

Training SAD, SCD or EMB will fail when dictionary as no 'annotation' key. 
However, prediction will still work... and MUST work because for evaluation campaigns, we don't have access to annotations on the test set.

Reply:
Ah, you want to allow the same subset (train, development, or test) with mixed audio files: some with annotations and some without? Can it really happen in practice? I thought that one subset is always of a certain type. If you have a training set in a database, it would either have all utterances annotated or not, but it would never have half annotated and half not.

Reply:
You are right: within a specific subset, it will be either 100% with annotation or 100% without annotation.

But it may happen that a 100% annotated file actually contains no annotation at all while still being correctly annotated (= does not contain any speech...).

Reply:
This is because `mdtm` files don't provide "non-speech" labels...

Reply:
Then, we still don't have a situation when the subset would have both files: the one with annotations in `rttm` (or `mdtm`) format and one a simple list file with just a single `uri` per line. Since it is either-or situation, I would think it's better to have just one file path parameter, so avoid confusion with all those different file paths. Three paths is a bit much.

Reply:
Yes. we do. Say we have two files (`file1` and `file2`) in the training set. 
* `file1` is a *regular* file with some speech turns.
* `file2` is a file with only non-speech.

If you only provide the `mdtm`, then you will only iterate on `file1` (because the `mdtm` won't refer to `file2` at all).
This means that all the non-speech data from `file2` won't be used when training speech activity detection. We don't want that, right?

Reply:
Ok, I didn't know such databases exist